pkg/kubevirt/actuator.go

```diff
@@ -0,0 +1,275 @@
+func (a *KubevirtActuator) Create(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	logger := a.getLogger(cr)
+	logger.Debug("running Create")
+	return a.sync(ctx, cr, logger)
+}
+
+// Update the credentials to the provided definition.
+func (a *KubevirtActuator) Update(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	logger := a.getLogger(cr)
+	logger.Debug("running Update")
+	return a.sync(ctx, cr, logger)
+}
+
+// Delete credentials
+func (a *KubevirtActuator) Delete(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	logger := a.getLogger(cr)
+	logger.Debug("running Delete")
+
+	existingSecret, err := a.getSecret(ctx, cr, logger)
+	if err != nil {
+		return err
+	}
+	if existingSecret != nil{
+		logger.Debug("Deleting existing secret")
+		if err = a.Client.Delete(ctx, existingSecret); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetCredentialsRootSecretLocation returns the namespace and name where the parent credentials secret is stored.
+func (a *KubevirtActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
+	return types.NamespacedName{Namespace: constants.CloudCredSecretNamespace, Name: constants.KubevirtCloudCredSecretName}
+}
+
+func (a *KubevirtActuator) sync(ctx context.Context, cr *minterv1.CredentialsRequest,  logger log.FieldLogger) error {
+	logger.Debug("running sync")
+
+	// get the secret data from the credentials request
+	kubevirtCredentialData, err := a.getCredentialsSecretData(ctx, logger)
+	if err != nil {
+		logger.WithError(err).Error("issue with cloud credentials secret")
+		return err
+	}
+
+	// get the existing secret in order to check if need to update or create a new
+	logger.Debug("provisioning secret")
+	existingSecret, err := a.getSecret(ctx, cr, logger)
+	if err != nil {
+		return err
+	}
+
+	// check if need to update or create a new one
+	if err = a.syncCredentialSecret(ctx, cr, &kubevirtCredentialData, existingSecret, logger); err != nil {
+		msg := "error creating/updating secret"
+		logger.WithError(err).Error(msg)
+		return &actuatoriface.ActuatorError{
+			ErrReason: minterv1.CredentialsProvisionFailure,
+			Message:   fmt.Sprintf("%v: %v", msg, err),
+		}
+	}
+	return nil
+}
+
+func (a *KubevirtActuator) getCredentialsSecretData(ctx context.Context, logger log.FieldLogger) ([]byte, error) {
+	// get the secret of the kubevirt credentials
+	kubevirtCredentialsSecret := &corev1.Secret{}
+	if err := a.Client.Get(ctx, a.GetCredentialsRootSecretLocation(), kubevirtCredentialsSecret); err != nil {
+		msg := "unable to fetch root cloud cred secret"
+		logger.WithError(err).Error(msg)
+		return nil, &actuatoriface.ActuatorError{
+			ErrReason: minterv1.CredentialsProvisionFailure,
+			Message:   fmt.Sprintf("%v: %v", msg, err),
+		}
+	}
+
+	// get the secret data - the infra kubeconfig
+	infraClusterKubeconfig, ok := kubevirtCredentialsSecret.Data[KubevirtCredentialsSecretKey]
+	if !ok {
+		return nil, errors.New("invalid mode")
```

Comment:
can we be more clear than "invalid mode". Maybe something like "credentials missing expected key".